tools/codegen/model.py

```diff
@@ -341,11 +335,34 @@ def from_yaml(ei: Dict[str, object], loc: 'Location') -> 'NativeFunction':
             "strictly subsumes the other.  If you wanted to provide an explicit autograd " \
             "implementation, specify CompositeExplicitAutograd; otherwise specify CompositeImplicitAutograd only"
 
+        if structured_delegate:
+            # Structured functions MUST have a dispatch table
+            is_abstract = True
+        else:
+            is_abstract = dispatch.keys() != {DispatchKey.CompositeImplicitAutograd}
+
+        has_composite_implicit_autograd_kernel = DispatchKey.CompositeImplicitAutograd in dispatch.keys()
+        has_composite_explicit_autograd_kernel = DispatchKey.CompositeExplicitAutograd in dispatch.keys()
+
+        # BackendMetadata is used to store any information about a NativeFunction that is backend dependent.
+        # The most obvious information is the kernel name, which usually contains the name of the backend in it for cpu/cuda.
+        # Why is 'structured' included? External backends (e.g. XLA) opt into which ops are structured
+        # independently of which in-tree ops are structured
+        backend_metadata = {k: {func.name: BackendMetadata(
+            kernel=v, structured=structured and is_structured_dispatch_key(k), external=False)} for k, v in dispatch.items()}
```

Comment:
Hmm, you sure about this? Isn't this only CPU/CUDA should be marked structured in the backend metadata?

Reply:
yep. Maybe I'm misunderstanding you, but isn't that what this is doing? I'm calling `is_structured_dispatch_key(k)`, which does that check.

So for all in-tree kernels, [is_structured_kernel] == [is_structured_op] & [is_structured_dispatch_key]

Reply:
missed the and!